PERF: Improve perf of to_datetime with ISO format

chris-b1 opened the pull request:
Closes #10178
Closes #8154

Using the example data from the issue:

```
In [2]: df = DataFrame({'date_text':["2015-05-18" for i in range(10**6)]})
```

Before:

```
In [3]: %timeit pd.to_datetime(df['date_text'],format="%Y-%m-%d", box=False).values.view('i8')/10**9
1 loops, best of 3: 3.14 s per loop

In [4]: %timeit pd.to_datetime(df['date_text'],infer_datetime_format=True,     box=False).values.view('i8')/10**9
1 loops, best of 3: 253 ms per lo
```

After:

```
In [6]: %timeit pd.to_datetime(df['date_text'],format="%Y-%m-%d", box=False).values.view('i8')/10**9
1 loops, best of 3: 217 ms per loop

In [7]: %timeit pd.to_datetime(df['date_text'],infer_datetime_format=True, box=False).values.view('i8')/10**9
1 loops, best of 3: 243 ms per loop
```

chris-b1 committed:
PERF: Improve perf of to_datetime with ISO format

jreback commented:
is their an existing vbench for this? if not pls add one (but make 1/10 the length)

chris-b1 commented:
I added a vbench based on existing one that inferred the iso860 format.

```
---------------------------------------------------------
Test name                                    |    #0    |
---------------------------------------------------------
timeseries_to_datetime_iso8601_format        |   4.9593 |
timeseries_to_datetime_YYYYMMDD              |  10.0046 |
timeseries_to_datetime_iso8601               |   4.9581 |
---------------------------------------------------------
Test name                                    |    #0    |
---------------------------------------------------------
```

The second line of this test is failing - is that really the desired behavior?  The comments reference #10154  - Wasn't the conclusion that `errors='ignore'` should always suppress?

    def test_day_not_in_month_coerce_false_ignore(self):
        self.assertEqual(to_datetime('2015-02-29', errors='ignore', coerce=False), '2015-02-29')
        self.assertRaises(ValueError, to_datetime, '2015-02-29', errors='ignore', format="%Y-%m-%d", coerce=False)
        self.assertRaises(ValueError, to_datetime, '2015-02-32', errors='ignore', format="%Y-%m-%d", coerce=False)
        self.assertRaises(ValueError, to_datetime, '2015-04-31', errors='ignore', format="%Y-%m-%d", coerce=False)

jreback commented:
The intention of that test behavior is that yes you will ignore errors (and you are not coercing), BUT, since you specified the format it is still an invalid date. 

so you bring up the case that it is doing different things depending on whether a format is passed or not.

rereading those comments in #10154. I think it is just a choice. Ok, so let's make them the same. I'll reverse myself and now say that all 4 of these should raise. @chris-b1 would you make that change.

cc @jorisvandenbossche 
cc @vincentdavis

jreback commented on pandas/tseries/tools.py:
```diff
@@ -300,17 +300,17 @@ def _convert_listlike(arg, box, format):
         if infer_datetime_format and format is None:
             format = _guess_datetime_format_for_array(arg, dayfirst=dayfirst)
```
maybe this should be:

```
if format is None:

    if infer_datetime_format:
           .......
    else:
          # check for the iso format
```

chris-b1 replied:
I think the iso format code needs to be outside the else - it should apply whether or not the format was inferred or specified?

chris-b1 commented:
%Y format with integer data doesn't work on iso fast code (causing Stata test problems) so I've excluded from this and added a new test.

I also changed the error handling in `to_datetime` to always raise on out of bound / unparse-able date unless `coerce=True`.  It ends up breaking quite a tests that expected the old behavior, just wanted to double check that's desired?  Thanks.

jorisvandenbossche commented:
Rereading the comments in the other issue: https://github.com/pydata/pandas/issues/10154#issuecomment-106484971 and the two comments below. Based on that, I would say we agreed that with `errors='ignore'`, no error should be raised (although it is not really clear what we decided). But I did not see that the actual PR then did something else. 
I think it is pretty clear: if you give `errors='ignore'`, it should not raise an error. But as I said over there, I don't really like this default behaviour.

chris-b1 commented:
I think that makes sense.   One corner case that comes up - what should happen if the data passed doesn't match the format, but can all be parsed.  E.g.

    pd.to_datetime(['2/28/2014', '2014-01-01'], format='%m/%d/%Y')
    pd.to_datetime(['2/28/2014', '2014-01-01'], format='%m-%d-%Y')

I'm not sure, but it's probably least surprising if it fails? That's what the latest commit does - adding an additional parameter to `tslib.array_to_datetime` to only parse iso8601 strings.  I wasn't too sure on the cleanest way to do that, but here's a first attempt that at least seems to work.

jreback commented on pandas/tseries/tests/test_timeseries.py:
```diff
@@ -4197,6 +4197,19 @@ def test_to_datetime_format_YYYYMMDD(self):
         assert_series_equal(result, expected)
 
+    def test_to_datetime_format_integer(self):
+        s = Series([2000, 2001, 2002])
```
add the issue number as a comment

jreback commented on pandas/tslib.pyx:
```diff
@@ -1908,6 +1909,12 @@ cpdef array_to_datetime(ndarray[object] values, raise_=False,
+                        if coerce:
+                            iresult[i] = iNaT
+                            continue
+                        raise TypeError
```
raise a ``ValueError``, but with a meaningul message (which you can then catch above). It is odd to convert to a ``TypeError``

jreback commented on pandas/tslib.pyx:
```diff
@@ -1970,10 +1977,15 @@ cpdef array_to_datetime(ndarray[object] values, raise_=False,
+                        _check_dts_bounds(&dts)
+                    else:
+                        if raise_:
+                            raise ValueError
```
need a meaningful message

jreback commented:
@chris-b1 there are already a lot of paths thru this code, so getting very hard to grok. Any simplication would be appreciated.

jorisvandenbossche commented:
After second thought, I am not sure anymore if this is a good idea (ignoring instead of raising), sorry ..

In any case, this is a backwards incompatible change, as it has been like that (that format raises instead of ignoring by default when it does not follow the format). It was just not correctly stated in the docstring. 
And if we do such a change, I would rather do it the other way around (change default of ignore to raise). But that is another discussion, and we should see if this change is not too breaking for people.

But maybe we can open a new issue for the discussion about the default behaviour, and keep this PR only for the perf improvement?

sinhrks commented:
I think ``_string_to_dts`` should handle ``require_iso8601`` kw for simplification. Raise ``DateParseError`` when it is ``True``, and ``ValueError`` otherwise. ``array_to_datetime``, can pass through ``DateParseError``, but catch ``ValueError``. 

  - https://github.com/pydata/pandas/blob/master/pandas/src/datetime.pxd#L127

For ``errors`` kw, I don't have clear opinion for now.

chris-b1 commented:
@sinhrks is it possible to raise/define an custom exception from a .pxd?  Sorry, running into the edge of my Cython knowledge.

jreback commented on pandas/tslib.pyx:
```diff
@@ -1908,6 +1909,17 @@ cpdef array_to_datetime(ndarray[object] values, raise_=False,
                 except ValueError:
+                    # if requiring iso8601 strings, skip trying other formats
+                    if require_iso8601:
+                        if coerce:
```
I think you can just use ``format is not None``. its passed (but not used) anyhow, rather than add another parameter?

sinhrks commented:
@chris-b1 Can use cdef (and convert it to normal python object when catch). But it does not simplify as I first thought... Please forget, and I'll do separately if I find better idea.

jreback commented:
@chris-b1 this looks ok. pls squash; ping when green.

I opened #10636 to address the changing of the default, which is prob a good idea for 0.17.0

@chris-b1 if you'd like to tackle would be gr8!

chris-b1 commented:
@jreback - should be good to merge.

jreback commented:
@chris-b1 thanks!

on #10636 we'll have to put the new behavior (that was slightly changed here).
